pysmt/solvers/bdd.py

```diff
@@ -0,0 +1,324 @@
+#
+# This file is part of pySMT.
+#
+#   Copyright 2014 Andrea Micheli and Marco Gario
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+import pycudd
+
+import pysmt.logics
+
+from pysmt import typing as types
+from pysmt.solvers.solver import Solver
+from pysmt.solvers.eager import EagerModel
+from pysmt.walkers import DagWalker
+from pysmt.decorators import clear_pending_pop, deprecated
+
+
+def assert_ddmanager(target):
+    """The global DdManager should match target.
+
+    pyCUDD uses a global DdManager. Therefore, we need to make sure
+    that we are operating on the right DdManager. This method enforces
+    that the global DdManager matches the one we are expecting.
+
+    Note: This solution might be stronger that what is actually
+    needed. In particular, it might be enough to simply set the
+    DdManager everytime time we operate on it.
+    """
+
+    assert pycudd.GetDefaultDdManager() == target, \
+        "Global DdManager does not match local DdManager. Multiple" +\
+        "DdManagers are currently not supported."
+
+
+class BddSolver(Solver):
+    LOGICS = [ pysmt.logics.QF_BOOL, pysmt.logics.BOOL ]
+
+    def __init__(self, environment, logic, options=None):
+        Solver.__init__(self,
+                        environment=environment,
+                        logic=logic,
+                        options=options)
+
+        self.mgr = environment.formula_manager
+        self.ddmanager = pycudd.DdManager()
+        self.ddmanager.SetDefault()
+        self.converter = BddConverter(environment=self.environment,
+                                      ddmanager=self.ddmanager)
+
+        self.assertions_stack = None
+        self.reset_assertions()
+
+        self.backtrack = []
+        self.latest_model = None
+
+    @clear_pending_pop
+    def reset_assertions(self):
+        true_formula = self.mgr.Bool(True)
+        self.assertions_stack = [(
+            true_formula,
+            self.converter.convert(true_formula))]
+
+    @clear_pending_pop
+    def declare_variable(self, var):
+        self.converter.declare_variable(var)
+
+    @clear_pending_pop
+    def add_assertion(self, formula, named=None):
+        assert_ddmanager(self.ddmanager)
+        bdd_expr = self.converter.convert(formula)
+        _, last_state = self.assertions_stack[-1]
+        new_state = last_state.And(bdd_expr)
```

Comment:
At stack level n you have the conjunction of all the levels i <= n...
* A comment is needed
* Why doing this computation here and not in the ```solve()``` method?